src/content.js

```diff
@@ -45,22 +45,77 @@ Function used to avoid multiple injection (cleaner than using an if?)
     var showTranslated = true;
     var orientation = 'horizontal-tb';
 
+    // Set to track successfully processed images to avoid reprocessing
+    const processedImages = new WeakSet();
+    // Set to track images currently being processed to avoid concurrent processing
+    const processingImages = new WeakSet();
+
+    // Helper function to check if a node is an image
+    function isImageNode(node) {
+        return node.nodeType === Node.ELEMENT_NODE && 
+               (node.tagName === 'IMG' || node.tagName === 'CANVAS');
+    }
+
+    // Helper function to process all images in a subtree
+    function processImagesInSubtree(root) {
+        if (isImageNode(root)) {
+            if (!processedImages.has(root)) {
+                debug('new image found in subtree', root);
+                processImage(root);
+            }
+            return;
+        }
+        
+        // Process all img and canvas elements in the subtree
+        const images = root.querySelectorAll ? root.querySelectorAll('img, canvas') : [];
+        images.forEach((img) => {
+            if (!processedImages.has(img)) {
+                debug('new image found via querySelectorAll', img);
+                processImage(img);
+            }
+        });
+    }
+
     var observer = new MutationObserver((mutations) => {
         mutations.forEach((mutation) => {
             if (mutation.type === 'childList') {
                 mutation.addedNodes.forEach((node) => {
-                    if (node.tagName === 'IMG' || node.tagName === 'CANVAS') {
-                        debug('image added', node);
-                        processImage(node);
+                    if (node.nodeType === Node.ELEMENT_NODE) {
+                        // Check if the node itself is an image
+                        if (isImageNode(node)) {
+                            debug('direct image added', node);
+                            processImage(node);
+                        } else {
+                            // Check for images within the added node
+                            processImagesInSubtree(node);
+                        }
```

Comment:
```suggestion
                        // Check for images within the added node
                        processImagesInSubtree(node);
```

Now that `processImagesInSubtree` is not recursively checking all nodes, but making use of the querySelector we can move the printing of `direct image added` inside and avoid doing the check twice